Sources/Logging/Logging.swift

```diff
@@ -653,55 +661,165 @@ extension Logger {
+    /// ``Logger/MetadataProvider`` which should be used by all loggers created in this process from here onwards.
+    ///
+    /// `bootstrapMetadataProvider` can be called at maximum once in any given program, calling it more than once will
+    /// lead to undefined behavior, most likely a crash.
+    ///
+    /// It is also possible to `bootstrap(metadataProvider:_:)` to configure both a metadata provider,
+    /// and log handler backend at the same time, in a single call.
+    ///
+    /// - Parameter metadataProvider: The metadata provider to be used by this logging system.
+    public static func bootstrapMetadataProvider(_ metadataProvider: Logger.MetadataProvider) {
+        self._metadataProviderFactory.replaceMetadataProvider(metadataProvider, validate: true)
+    }
+
+    // for our testing we want to allow multiple bootstrapping
     internal static func bootstrapInternal(_ factory: @escaping (String) -> LogHandler) {
+        self._metadataProviderFactory.replaceMetadataProvider(.noop, validate: false) // TODO: fixme remove this and fix tests
+        self._factory.replaceFactory({ label, _ in
+            factory(label)
+        }, validate: false)
+    }
+
+    // for our testing we want to allow multiple bootstraping
+    internal static func bootstrapInternal(_ factory: @escaping (String, Logger.MetadataProvider) -> LogHandler) {
         self._factory.replaceFactory(factory, validate: false)
     }
 
-    fileprivate static var factory: (String) -> LogHandler {
-        return self._factory.underlying
+    internal static func bootstrapInternal(metadataProvider: Logger.MetadataProvider) {
+        self._metadataProviderFactory.replaceMetadataProvider(metadataProvider, validate: false)
+    }
+
+    internal static func bootstrapInternal(metadataProvider: Logger.MetadataProvider,
+                                           _ factory: @escaping (String, Logger.MetadataProvider) -> LogHandler) {
+        self._metadataProviderFactory.replaceMetadataProvider(metadataProvider, validate: false)
+        self._factory.replaceFactory(factory, validate: false)
+    }
+
+    fileprivate static var factory: (String, Logger.MetadataProvider) -> LogHandler {
+        return { label, metadataProvider in
+            var handler = self._factory.underlying(label, metadataProvider)
+            handler.metadataProvider = metadataProvider // TODO: necessary or not?
+            return handler
+        }
+    }
+
+    /// System wide ``Logger/MetadataProvider`` that was configured during the logging system's `bootstrap`.
+    internal static var metadataProvider: Logger.MetadataProvider {
+        return self._metadataProviderFactory.metadataProvider
     }
 
     private final class FactoryBox {
         private let lock = ReadWriteLock()
-        fileprivate var _underlying: (String) -> LogHandler
+        fileprivate var _underlying: (String, Logger.MetadataProvider) -> LogHandler
```

Comment:
nit: add labels

Reply:
Ok, though they'll be pretty boring in the sense that not usable at call sites: `fileprivate var _underlying: (_ label: String, _ provider: Logger.MetadataProvider)`; swift does not allow for `_underlying: (label: String, provider: Logger.MetadataProvider)` :-/

Reply:
(done)